File: gptcomet/llms/gemini.py
from typing import Any, Optional

from gptcomet.log import logger
from gptcomet.utils import console

from .base import BaseLLM


class GeminiLLM(BaseLLM):
    """Gemini LLM provider implementation."""

    def __init__(self, config: dict[str, Any]):
        super().__init__(config)

        self.api_base = self.api_base or "https://generativelanguage.googleapis.com/v1beta/models"
        self.model = self.model or "gemini-pro"
        self.completion_path = self.completion_path or "generateContent"
        self.answer_path = self.answer_path or "candidates.0.content.parts.0.text"

    def format_messages(
        self, message: str, history: Optional[list[dict[str, str]]] = None
    ) -> dict[str, Any]:
        """Format messages for Gemini API."""
        contents = []

        if history:
            for msg in history:
                role = "model" if msg["role"] == "assistant" else "user"
                contents.append({"role": role, "parts": [{"text": msg["content"]}]})

        contents.append({"role": "user", "parts": [{"text": message}]})

        payload = {
            "contents": contents,
            "generationConfig": {
                "maxOutputTokens": self.max_tokens,
            },
        }
        if self.temperature is not None:
            payload["generationConfig"]["temperature"] = float(self.temperature)

        if self.top_p is not None:
            payload["generationConfig"]["topP"] = float(self.top_p)

        if self.frequency_penalty is not None:
            payload["generationConfig"]["frequencyPenalty"] = float(self.frequency_penalty)

        if self.presence_penalty is not None:
            payload["generationConfig"]["presencePenalty"] = float(self.presence_penalty)

        return payload

    def build_url(self) -> str:
        """Build the API URL."""
        return f"{self.api_base}/{self.model}:generateContent?key={self.api_key}"

    def build_headers(self) -> dict[str, str]:
        """Build request headers."""
        return {
            "Content-Type": "application/json",
        }

    def make_request(
        self, message: str, history: Optional[list[dict[str, str]]] = None, **kwargs
    ) -> str:
        """Make a request to the API."""
        url = self.build_url()
        headers = self.build_headers()
        payload = self.format_messages(message, history)
        logger.debug("Sending request...")

        with self.managed_session() as session:
            response = session.post(url, json=payload, headers=headers, timeout=self.timeout)
            logger.debug(f"Request payload: {payload}")
            logger.debug(f"Response: {response.json()}")
            response.raise_for_status()
            data = response.json()
            usage = self.get_usage(data)
            if usage:
                console.print(usage)
            return self.parse_response(data)

    def get_usage(self, data: dict[str, Any]) -> Optional[str]:
        """Print usage information for the provider."""
        usage = data.get("usageMetadata")
        if not usage:
            return None
        else:
            return (
                f"Token usage: promptTokenCount: {usage.get('promptTokenCount')}, "
                f"candidatesTokenCount: {usage.get('candidatesTokenCount')}, "
                f"totalTokenCount: {usage.get('totalTokenCount')}"
            )

    @classmethod
    def get_required_config(cls) -> dict[str, tuple[str, str]]:
        return {
            "api_base": (
                "https://generativelanguage.googleapis.com/v1beta/models",
                "Enter Gemini API base",
            ),
            "model": ("gemini-pro", "Enter Gemini model"),
            "api_key": ("", "Enter Gemini API key"),
            "max_tokens": ("1024", "Enter max tokens"),
        }

File: internal/llm/gemini.go
package llm

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"

	"github.com/tidwall/gjson"

	"github.com/belingud/gptcomet/internal/debug"
	"github.com/belingud/gptcomet/pkg/config"
	"github.com/belingud/gptcomet/pkg/types"
)

// GeminiLLM implements the LLM interface for Gemini
type GeminiLLM struct {
	*BaseLLM
}

// NewGeminiLLM creates a new GeminiLLM
func NewGeminiLLM(config *types.ClientConfig) *GeminiLLM {
	if config.APIBase == "" {
		config.APIBase = "https://generativelanguage.googleapis.com/v1beta/models"
	}
	if config.Model == "" {
		config.Model = "gemini-1.5-flash"
	}
	if config.CompletionPath == "" {
		config.CompletionPath = "generateContent"
	}
	if config.AnswerPath == "" {
		config.AnswerPath = "candidates.0.content.parts.0.text"
	}

	return &GeminiLLM{
		BaseLLM: NewBaseLLM(config),
	}
}

func (g *GeminiLLM) Name() string {
	return "gemini"
}

// GetRequiredConfig returns provider-specific configuration requirements
func (g *GeminiLLM) GetRequiredConfig() map[string]config.ConfigRequirement {
	return map[string]config.ConfigRequirement{
		"api_base": {
			DefaultValue:  "https://generativelanguage.googleapis.com/v1beta/models",
			PromptMessage: "Enter Gemini API base",
		},
		"model": {
			DefaultValue:  "gemini-1.5-flash",
			PromptMessage: "Enter model name",
		},
		"api_key": {
			DefaultValue:  "",
			PromptMessage: "Enter API key",
		},
		"max_tokens": {
			DefaultValue:  "1024",
			PromptMessage: "Enter max tokens",
		},
	}
}

// FormatMessages formats messages for Gemini API
func (g *GeminiLLM) FormatMessages(message string, history []types.Message) (interface{}, error) {
	var contents []map[string]interface{}

	contents = append(contents, map[string]interface{}{
		"role":  "user",
		"parts": []map[string]string{{"text": message}},
	})

	payload := map[string]interface{}{
		"contents": contents,
		"generationConfig": map[string]interface{}{
			"maxOutputTokens": g.Config.MaxTokens,
		},
	}

	if g.Config.Temperature > 0 {
		payload["generationConfig"].(map[string]interface{})["temperature"] = g.Config.Temperature
	}
	if g.Config.TopP > 0 {
		payload["generationConfig"].(map[string]interface{})["topP"] = g.Config.TopP
	}
	if g.Config.FrequencyPenalty > 0 {
		payload["generationConfig"].(map[string]interface{})["frequencyPenalty"] = g.Config.FrequencyPenalty
	}
	if g.Config.PresencePenalty > 0 {
		payload["generationConfig"].(map[string]interface{})["presencePenalty"] = g.Config.PresencePenalty
	}
	debug.Printf("generationConfig: %v", payload["generationConfig"])

	return payload, nil
}

// BuildURL builds the API URL
func (g *GeminiLLM) BuildURL() string {
	return fmt.Sprintf("%s/%s:generateContent?key=%s", strings.TrimSuffix(g.Config.APIBase, "/"), g.Config.Model, g.Config.APIKey)
}

// BuildHeaders builds request headers
func (g *GeminiLLM) BuildHeaders() map[string]string {
	return map[string]string{
		"Content-Type": "application/json",
	}
}

// GetUsage returns usage information for the provider
func (g *GeminiLLM) GetUsage(data []byte) (string, error) {
	usage := gjson.GetBytes(data, "usageMetadata")
	if !usage.Exists() {
		return "", nil
	}

	return fmt.Sprintf(
		"Token usage> promptTokenCount: %d, candidatesTokenCount: %d, totalTokenCount: %d",
		usage.Get("promptTokenCount").Int(),
		usage.Get("candidatesTokenCount").Int(),
		usage.Get("totalTokenCount").Int(),
	), nil
}

// MakeRequest makes a request to the API
func (g *GeminiLLM) MakeRequest(ctx context.Context, client *http.Client, message string, history []types.Message) (string, error) {
	url := g.BuildURL()
	headers := g.BuildHeaders()
	payload, err := g.FormatMessages(message, history)
	if err != nil {
		return "", fmt.Errorf("failed to format messages: %w", err)
	}

	debug.Printf("Sending request...")

	reqBody, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(reqBody))
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read response: %s", err)
	}

	debug.Printf("Response: %s", string(respBody))

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(respBody))
	}

	usage, err := g.GetUsage(respBody)
	if err != nil {
		return "", fmt.Errorf("failed to get usage: %w", err)
	}
	if usage != "" {
		fmt.Printf("%s\n", usage)
	}

	return g.ParseResponse(respBody)
}
